diff --git a/arch/arm/kernel/vmlinux.lds.S b/arch/arm/kernel/vmlinux.lds.S
--- a/arch/arm/kernel/vmlinux.lds.S
+++ b/arch/arm/kernel/vmlinux.lds.S
@@ -215,4 +215,9 @@ 
 		__pv_table_end = .;
	}
+	.init.hdf_table : {
+		_hdf_drivers_start = .;
+		*(.hdf.driver)
+		_hdf_drivers_end = .;
+	}
	.init.data : {
		INIT_DATA

diff --git a/drivers/Makefile b/drivers/Makefile
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -181,2 +181,6 @@ 
 obj-$(CONFIG_ARCH_HISI_BVT)	+= hisilicon/
 obj-$(CONFIG_HI_VDMA_V100) += hi_vdmav100/
+#ifeq ($(CONFIG_DRIVERS_HDF),y)
+obj-y	+= hdf/hdf_adapt.o
+#endif
+#error stop

diff --git a/drivers/hdf/hdf_adapt.c b/drivers/hdf/hdf_adapt.c
new file mode 100644
--- /dev/null
+++ b/drivers/hdf/hdf_adapt.c
@@ -0,0 +0,8 @@ 
+#include <stddef.h>
+#include <linux/module.h>
+
+extern size_t _hdf_drivers_start;
+EXPORT_SYMBOL(_hdf_drivers_start);
+
+extern size_t _hdf_drivers_end;
+EXPORT_SYMBOL(_hdf_drivers_end);
